### PR TITLE
Create attention layers with external kv cache and prepare an example.

### PR DESCRIPTION
**Pull request by haozha111**
This PR contains following changes:
1) Fork the current attention code into layers/experimental/attention.py, which supports external KV cache style (for GPU work). This will avoid polluting the current internal-kv attention implementation.
2) Put SDPA functions into separate files to allow code reuse.
3) Prepare a toy model w/ 2 transformer blocks, using external kv cache layers.

Note conversion to TFL fails because tfl's dynamic update slice doesn't support int64 index, will work on a fix later.

BUG=b/338453350

### CODE REVIEW

File: ai_edge_torch/generative/layers/experimental/attention.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+
+# CausalSelfAttention which can support MHQ, MQA or GQA.
+class CausalSelfAttention(nn.Module):
+
+  def __init__(
+      self,
+      dim: int,
+      config: cfg.AttentionConfig,
+      kv_cache_max: int,
```
Comment (hheydary) [inline ai_edge_torch/generative/layers/experimental/attention.py:103]: This argument does not seem to be used anywhere.

File: ai_edge_torch/generative/examples/test_models/toy_model_with_external_kv_cache.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  ) -> (torch.Tensor, torch.Tensor, torch.Tensor):
+    x = self.tok_embedding(idx)
+    cos, sin = self.rope_cache
+    cos = cos.index_select(0, input_pos)
+    sin = sin.index_select(0, input_pos)
+    mask = self.mask_cache.index_select(2, input_pos)
+    mask = mask[:, :, :, : self.config.max_seq_len]
+
+    for i, block in enumerate(self.transformer_blocks):
+      input_k = torch.clone(k_caches[i])
```
Comment (hheydary) [inline ai_edge_torch/generative/examples/test_models/toy_model_with_external_kv_cache.py:76]: Why are we cloning these tensors?

Comment (haozha111) [reply ai_edge_torch/generative/examples/test_models/toy_model_with_external_kv_cache.py:76]: good catch, we don't need those. originally i thought this was needed to walk-around a dynamo export issue, but trying again seems it works even without this hack.

File: ai_edge_torch/generative/layers/experimental/attention.py (hunk 1 above)
Comment (haozha111) [reply ai_edge_torch/generative/layers/experimental/attention.py:103]: removed.

File: ai_edge_torch/generative/layers/experimental/attention.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      x_norm = self.pre_atten_norm(x)
+      attn_out, kv_cache = self.atten_func(x_norm, rope, mask, input_pos, kv_cache)
+      x = x + attn_out
+      x_norm = self.pre_ff_norm(x)
+      output = x + self.ff(x_norm)
+
+    return output, kv_cache
+
+
+# CausalSelfAttention which can support MHQ, MQA or GQA.
```
Comment (hheydary) [inline ai_edge_torch/generative/layers/experimental/attention.py:95]: This could be added as class docstring instead of inline comment.

**Review by hheydary — DISMISSED** [review]

File: ai_edge_torch/generative/layers/experimental/attention.py (hunk 3 above)
Comment (haozha111) [reply ai_edge_torch/generative/layers/experimental/attention.py]: cool, original layer updated as well. now we need to keep both in sync to ensure a smooth migration.

**Review by hheydary — APPROVED** [review]

**Pull request merged**